spec/history.cddl

```diff
@@ -0,0 +1,39 @@
+canister-history = tagged<{
+  total_num_changes : unsigned
+  change_entries : [* canister-change-entry]
+}>
+
+canister-change-entry = {
+  timestamp : timestamp
+  origin : canister-change-origin
+  change : canister-change
+}
+
+canister-change-origin = {
+  from : "user"
+  user_id : principal
+  //
+  from : "canister"
+  canister_id : principal
+  canister_history_index : unsigned
```

Comment:
I think this is more fundamental than it may initially appear: What we are effectively saying here is that a canister is not purely identified by its canister id, but by the canister id _and_ the version. Then if we're consistent, we may ask questions like:
- Shouldn't we also expose this in the `caller` in inter-canister calls?
- There was once the idea of sending a message conditionally on the canister running a certain hash, shouldn't we allow sending a message conditionally on the target canister being at a certain version?
- Shouldn't the certified variables of a canister specify the version of the canister that produced them?

Of course we don't have to do all of this at once, but I think the change is (also witnessed by the fact that we'd actually have to change the message routing protocol) more fundamental than it initially seems.

Would appreciate other thoughts – so let me also ping @nomeata !

Reply:
I'd say the canister evolution is an internal to the canister, and something that, at least in general, should be of no concern to other parties.

Note that the universal canister can emulator all (or most) those changes without a history change. I'd hide whether a state change affects just the data or also the code. At least initially.

Reply:
@nomeata you were already proposing something similar [here](https://github.com/dfinity/interface-spec/pull/19/files#diff-4de6dbd4bce188297247b78b84639ec28500b6e90a783241539e95c9e33fe4d0R2109)

Reply:
Ah, yes: When handling a callback the canister should know the generation that sent the callback, for the reasons explained in #19. But this is still internal (like the `env`), and not visible to the other canister. It seems that Björn is pondering whether to expose the generation to the other side.

I wonder if `canister_history_index` can be used as in #19, despite the fact that it can increase during the lifetime of one instance, or if we’d still need a separate instance generation counter. But I think it works: the canister can store the current index in `canister_init`, and can then reliably distinguish the current generation (history index ≥ the stored one) from old ones (history index < the stored one).

Reply:
I think you're right that exposing `canister_history_index` via System API (e.g., `ic0.canister_history_index`) would enable to distinguish callbacks from the current generation from callbacks from past ones. Still, this would go beyond the scope of this PR.

My main motivation for introducing `canister_history_index` is the ability to implement the following notion: Given a canister C, identify all canisters D such that the following holds: for every canister E that is a (reflexive and transitive) controller of C, we have that E has been (reflexively and transitively) deployed by D.

If this property holds, then the canister D has “full” control over the current and future deployments of C.

Hence, it enables checking whether a canister (C) is managed by some DAO, e.g., SNS governance (canister D).

Here, checking if "E has been (reflexively and transitively) deployed by D" requires `canister_history_index` or sufficient gaps between the time-stamps of the individual deployments in the chain between D and E. In the latter case, however, it is not clear what is "sufficient" and, according to @daniel-wong-dfinity-org from the NNS team, "not relying on time for synchronization is highly preferred".

Reply:
I agree that having the history index here is sensible.

Reply:
This is a  useful discussion, and I think I see a clear path forward here that generalizes beyond the imminent need.

Multiple times we noticed a need to somehow identify and discern different points in the “canister-local timeline”:

* In the context of #19 we want to recognize canister generations (changes to the canister _code_).

* Here, the SNN want to recognize canister configurations (changes to the canister code and/or certain system metadata).

* application developers face difficulties with making sense of parallel update calls, or update calls and query calls, because of asynchronicity and potentially reaching different replica. They would dearly have some form of canister state counter included in the response to learn their real ordering. (Changes to the canister user state)

I have no reason to believe that this is the end, and likely more use cases will occur.

We could add three different concepts as we try to address these issues. But that's redundant and short-sighted design, because all these use cases are also solvable with a _finer_ counter. 

So my recommendation is to finally add the “canister state counter”. A simple number, bumped whenever anything about the canister state in the broad sense (code, system state, user state) changes. One counter to rule them all.

For example then it suffices for the SNN use case it all history entries include the canister state counter, and also the canister origin (as shown above): it can tell which interval it falls in.

This also nicely avoids the problem that we don't have a good name for CanisterHistoryIndex :-)

Reply:
Agreed! This suggestion sounds really useful!

Reply:
Yes, that looks really good! Just to throw in another possible application of this: For "secure" queries (meaning: ask multiple replicas, aggregate the result) we have to make sure we execute the query at the right state. So far we only have block height (not nice, exposes protocol internals) or subnet time (not nice since not guaranteed to be strictly increasing). The canister state counter es *exactly* the right tool for that job as well.

Reply:
Exactly!

Reply:
I’d create one PR to introduce the counter concept itself, and then separate PRs (such as this one) to expose/use the counter in various places

(Maybe the plain system API exposure to the canister can be in the first PR, but I would not lump it together with unrelated big changes like the Canister History log.)

Reply:
Here's the [PR](https://github.com/dfinity/interface-spec/pull/106).